feat: add WebRTC transport

melekes opened the pull request:
Refs https://github.com/paritytech/smoldot/issues/1712
Impl https://github.com/libp2p/rust-libp2p/pull/2622
polkadot companion: https://github.com/paritytech/polkadot/pull/6529
cumulus companion: https://github.com/paritytech/cumulus/pull/2076
Closes paritytech/polkadot-sdk#547

- In `dev` mode the node will advertise and accept `/webrtc` connections;
- The initial plan was to enable it for non-validators as well, but ultimately we've decided to postpone this until the majority of nodes & downstream users (parachains) upgrade to `libp2p-identify` 0.41.1 (see https://github.com/paritytech/substrate/pull/12529#discussion_r1217863946);
- The transport will generate and store the certificate, which is required for WebRTC identity, in
base dir. In the future, when [a new version of `ring` library is released](https://github.com/briansmith/ring/issues/1504), the certificate will be
deterministically derived from the node's peer ID.

melekes committed:
client: add WebRTC transport

Refs https://github.com/paritytech/smoldot/issues/1712
Impl https://github.com/libp2p/rust-libp2p/pull/2622

- WebRTC transport is enabled for non-validators and developers by default.
- The transport will generate and store the certificate, which is required for WebRTC identity, in
base dir. In the future, when a new version of `ring` library is released, the certificate will be
deterministically derived from the node's peer ID.

tomaka commented on client/network/src/transport.rs:
```diff
@@ -116,8 +118,19 @@ pub fn build_transport(
 
-	(transport, bandwidth)
+	if !memory_only {
+		let webrtc_transport = WebRTCTransport::new(keypair);
```
I don't actually know if this is the appropriate place, but the WebRTC certificate should be saved and loaded from disk. We already do this with the `keypair`, you should be able to follow the trail and find where we do it.

tomaka commented:
This is not in the diff, but I would suggest to copy-paste the following line:

https://github.com/tomaka/polkadot/blob/07e1b6d72747f3cde49e85e4ff68ff158daebe16/client/network/src/discovery.rs#L662-L666

And put it here as well:

https://github.com/tomaka/polkadot/blob/07e1b6d72747f3cde49e85e4ff68ff158daebe16/client/network/src/discovery.rs#L697

(except without the word `external`)

Otherwise, it's quite annoying to figure out what your `certhash` is.

melekes commented on client/cli/src/params/network_params.rs:
```diff
@@ -182,7 +186,26 @@ impl NetworkParams {
+
+			// Enable WebRTC for non-validators and developers by default.
+			if !is_validator || is_dev {
+				let port = self.webrtc_port.unwrap_or(default_listen_port + 1);
```
Thought about this too, but having two identical ports might lead to confusion in the future (when debugging something).

tomaka replied:
I think that automatically doing `default_listen_port + 1` would lead to even more confusion

tomaka commented on client/network/src/config.rs:
```diff
@@ -445,6 +461,67 @@ where
+#[derive(Clone, Debug)]
+pub enum WebRTCConfig {
+	/// A WebRTC certificate.
+	Certificate(Option<PathBuf>),
```
```suggestion
	FromDisk(Option<PathBuf>),
```

?

If you name this `Certificate`, I don't really see what the other variants would be

melekes commented on Cargo.toml:
```diff
@@ -310,3 +310,6 @@ inherits = "release"
 codegen-units = 1
+
+[patch.crates-io]
+libp2p = { version = "0.50.0", git = "https://github.com/libp2p/rust-libp2p.git", branch = "master" }
```
this needs to be addressed before merging

tomaka commented on client/network/src/config.rs:
```diff
@@ -445,6 +461,67 @@ where
+/// The configuration of a node's WebRTC certificate, describing how it is obtained.
+#[derive(Clone, Debug)]
+pub enum WebRTCConfig {
+	/// A WebRTC certificate.
+	File(Option<PathBuf>),
```
```suggestion
	/// Certificate stored on disk.
	/// A new certificate is randomly generated and written there if the file doesn't exist.
	File(PathBuf),
	Temporary,
```

tomaka commented on client/network/src/config.rs:
```diff
@@ -445,6 +461,67 @@ where
+	File(Option<PathBuf>),
+}
+
+impl Default for WebRTCConfig {
+	fn default() -> WebRTCConfig {
+		Self::File(None)
+	}
+}
```
I'm not sure that we should provide a default. You are very much supposed to load a certificate from disk (at least until the deterministic certificate system lands).

melekes replied:
I've added `Ephemeral` option for tests. Let me know if you think we should use `File` option in all tests too.

tomaka commented on client/network/src/service.rs:
```diff
@@ -381,7 +384,19 @@ where
 			};
 
-			let mut builder = SwarmBuilder::new(transport, behaviour, local_peer_id)
+			let builder = if let Some(spawner) = params.executor {
+				struct SpawnImpl<F>(F);
+				impl<F: Fn(Pin<Box<dyn Future<Output = ()> + Send>>)> Executor for SpawnImpl<F> {
+					fn exec(&self, f: Pin<Box<dyn Future<Output = ()> + Send>>) {
+						(self.0)(f)
+					}
+				}
+				SwarmBuilder::with_executor(transport, behaviour, local_peer_id, SpawnImpl(spawner))
+			} else {
+				let tp = ThreadPoolBuilder::new().name_prefix("libp2p-swarm-task-").create()?;
+				SwarmBuilder::with_executor(transport, behaviour, local_peer_id, tp)
+			};
```
Is this related to WebRTC? If not, I'd prefer if this was a separate PR.

kpp replied:
This is what we need to upgrade to libp2p 0.50. I believe it would be nice to create a separate PR so we can share the code.

melekes replied:
> Is this related to WebRTC? If not, I'd prefer if this was a separate PR.

https://github.com/paritytech/substrate/pull/12734

melekes replied:
But again, if I remove the above code, the PR will no longer compile 😕

altonen replied:
What if you set the base branch of this branch to point to the libp2p upgrade branch?

tomaka commented on client/network/src/discovery.rs:
```diff
@@ -624,6 +624,12 @@ impl NetworkBehaviour for DiscoveryBehaviour {
 	}
 
 	fn inject_new_listen_addr(&mut self, id: ListenerId, addr: &Multiaddr) {
+		info!(
+			target: "sub-libp2p",
+			"🔍 Discovered new address for our node: {}",
+			addr,
+		);
+
```
Same remark: this should probably be a separate PR

tomaka reviewed (changes requested):
It's still unclear to me what to do w.r.t. the bandwidth
Everything would be way easier if webrtc-rs simply kept track of how many bytes have been sent and received in total, which would let us calculate the bandwidth, it really doesn't seem like a difficult change to me

tomaka commented on client/cli/src/params/webrtc_certificate_params.rs:
```diff
@@ -0,0 +1,118 @@
+			return Ok(WebRTCConfig::Ephemeral)
+		}
+
+		let filename = self
+			.webrtc_certificate_file
+			.clone()
+			.unwrap_or_else(|| net_config_dir.join(WEBRTC_CERTIFICATE_FILENAME));
```
The default WebRTC file should be put within the path indicated by the `base_dir` CLI option.

tomaka commented on client/cli/src/params/webrtc_certificate_params.rs:
```diff
@@ -0,0 +1,118 @@
+	/// Create a `WebRTCConfig` from the given `WebRTCCertificateParams` in the context
+	/// of an optional network config storage directory.
+	pub fn webrtc_certificate(&self, net_config_dir: &PathBuf) -> error::Result<WebRTCConfig> {
+		if let Some(true) = self.webrtc_certificate_ephemeral {
+			return Ok(WebRTCConfig::Ephemeral)
+		}
```
The `--tmp` CLI flag should also trigger `Ephemeral`

bkchr replied:
As long as the WebRTC file will be put into `base_dir`, it will be ephemeral automatically. `--tmp` makes the `base_dir` a temporary folder that will be deleted when the node exists.

melekes committed:
upgrade libp2p to 0.50.0

melekes committed:
on_swarm_event and on_connection_handler_event

melekes committed:
replace `Swarm::new` with `Swarm::with_threadpool_executor`

melekes committed:
on_swarm_event and on_connection_handler_event part 2

melekes committed:
on_swarm_event and on_connection_handler_event part 3

melekes committed:
on_swarm_event and on_connection_handler_event part 4

melekes committed:
update libp2p

melekes commented:
sorry had to force push to do https://github.com/paritytech/substrate/pull/12529#discussion_r1027656190

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2071272

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2071270

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2071321

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2071308

melekes committed:
put webrtc certificate within base, not network config directory

https://github.com/paritytech/substrate/pull/12529#pullrequestreview-1185732327

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2071361

tomaka commented:
> sorry had to force push to do [#12529 (comment)](https://github.com/paritytech/substrate/pull/12529#discussion_r1027656190)

It doesn't matter now, but you didn't have to do that. We squash commits anyway, so whether you're based on top of another branch doesn't matter. The point was just to merge this in two times: first the libp2p changes, then the WebRTC changes.

melekes commented:
> so whether you're based on top of another branch doesn't matter.

well it matters for me because the amount of conflicts reduces dramatically when you rebase 1 commit vs 22 😄

melekes committed:
Merge branch 'master' into anton/upgrade-libp2p-to-0.50.0

melekes committed:
libp2p 0.50.0

melekes committed:
rename OutboundQueryCompleted to OutboundQueryProgressed

refs https://github.com/libp2p/rust-libp2p/pull/2712

melekes committed:
remove unused var

melekes committed:
accumulate outbound_query_records until query is finished

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2092840

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2092828

melekes commented:
can we ship first version without bandwidth?

tomaka commented:
Unless someone comments here or on the forum, let's ship without the bandwidth, provided an issue is opened to keep track of this.

melekes committed:
Merge branch 'master' into anton/webrtc-transport

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2243562

melekes committed:
fixes after the merge

melekes commented:
Check labels job was failing because this PR didn't have `D` label, so I added one. Not sure it's needed though

melekes committed:
fix docs

dmitry-markin reviewed (approved):
LGTM!

melekes commented on client/network/src/transport.rs:
```diff
@@ -116,8 +119,19 @@ pub fn build_transport(
+	if !memory_only {
+		let webrtc_transport = WebRTCTransport::new(keypair, webrtc_cert);
+
+		let transport = OrTransport::new(webrtc_transport, transport)
+			.map(|either_output, _| match either_output {
+				EitherOutput::First((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
+				EitherOutput::Second((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
+			})
+			.boxed();
```
> A little bit offtopic, but what is the difference between OrTransport and EitherTransport in libp2p? Wasn't able to understand it from the docs.

`OrTransport` will first try 1st transport. If that fails with error (say `Multiaddr` is not supported), `OrTransport` will try the 2nd transport.

`EitherTransport` acts more like a https://docs.rs/either/latest/either/enum.Either.html, where you have two cases and `EitherTransport` will use one that maches (either Left or Right).

nazar-pc replied:
So this update makes WebRTC transport preferable over plan TCP? I think it should be the opposite.

melekes replied:
> Wrapping muxer into StreamMuxerBox is not done above when combining TCP transport and WebSocket transport. Why is it needed here when combining with WebRTCTransport?

General info: TCP and WebSocket transports do not do multiplexing. In that way they can be easily combined. Yamux multiplexing is layered on top, transforming them into a stream muxer. In contrast, `WebRTCTransport` and `QUICTransport` both have streams and do "multiplexing in-house". So they are already muxers.

Answering your question, I had to wrap both into `StreamMuxerBox` so `Either` branches have the same type. Otherwise, you'd get:

```
    = note: expected tuple `(PeerId, libp2p::libp2p_webrtc::tokio::Connection)`
               found tuple `(PeerId, EitherOutput<Yamux<libp2p::libp2p_yamux::Incoming<multistream_select::negotiated::Negotiated<NoiseOutput<multistream_select::negotiated::Negotiated<BandwidthConnecLogging<EitherOutput<EitherOutput<EitherOutput<rw_stream_sink::RwStreamSink<BytesConnection<libp2p::libp2p_tcp::tokio::TcpStream>>, libp2p::libp2p_tcp::tokio::TcpStream>, EitherOutput<rw_stream_sink::RwStreamSink<BytesConnection<libp2p::libp2p_tcp::tokio::TcpStream>>, libp2p::libp2p_tcp::tokio::TcpStream>>, rw_stream_sink::RwStreamSink<Chan>>>>>>>>, libp2p::libp2p_mplex::Multiplex<multistream_select::negotiated::Negotiated<NoiseOutput<multistream_select::negotiated::Negotiated<BandwidthConnecLogging<EitherOutput<EitherOutput<EitherOutput<rw_stream_sink::RwStreamSink<BytesConnection<libp2p::libp2p_tcp::tokio::TcpStream>>, libp2p::libp2p_tcp::tokio::TcpStream>, EitherOutput<rw_stream_sink::RwStreamSink<BytesConnection<libp2p::libp2p_tcp::tokio::TcpStream>>, libp2p::libp2p_tcp::tokio::TcpStream>>, rw_stream_sink::RwStreamSink<Chan>>>>>>>>)`
```

melekes replied:
@nazar-pc good catch 👍 addressed in https://github.com/paritytech/substrate/pull/12529/commits/0db37d09f470f44805b19e981b5833c0e090898d

melekes committed:
better vars names in tests

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2249169

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2249170

melekes commented:
Hmmm... tests are failing with:

```
--- STDERR:              frame-support-test::derive_no_bound test_enum ---
thread 'test_enum' panicked at 'assertion failed: `(left == right)`
  left: `"Enum::VariantUnnamed(1, 2, 3, PhantomData<(derive_no_bound::ImplNone, derive_no_bound::ImplNone)>)"`,
 right: `"Enum::VariantUnnamed(1, 2, 3, PhantomData)"`', frame/support/test/tests/derive_no_bound.rs:212:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panicking.rs:65:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
   4: core::ops::function::FnOnce::call_once
   5: core::ops::function::FnOnce::call_once
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/ops/function.rs:251:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

melekes committed:
Merge branch 'master' into anton/webrtc-transport

melekes committed:
always give preference to tcp transport

https://github.com/paritytech/substrate/pull/12529#discussion_r1066583014

melekes committed:
Merge branch 'master' into anton/webrtc-transport

burdges commented:
Just curious, how do we authenticate the DTLS cert here?  ToFU?

mxinden commented:
@burdges unfortunately the browser is very restrictive when it comes to WebRTC. We can not inject anything in the DTLS handshake itself. Instead we run an additional Noise handshake, but don't double encrypt thereafter. See https://github.com/libp2p/specs/tree/master/webrtc#connection-security for details.

burdges commented:
We should get someone to send a PR to firefox it seems.

melekes committed:
Merge branch 'master' into anton/webrtc-transport

melekes committed:
use udp protocol, not tcp for webrtc

melekes committed:
copy warp-sync zombienet test

melekes committed:
rename test files

michalkucharczyk commented on scripts/ci/gitlab/pipeline/zombienet.yml:
```diff
@@ -50,3 +50,11 @@ zombienet-0001-basic-warp-sync:
+  script:
+    - /home/nonroot/zombie-net/scripts/ci/run-test-env-manager.sh
+        --github-remote-dir="${GH_DIR}/0004-webrtc"
+        --test="test-warp-sync.zndsl"
```
```suggestion
        --test="test-webrtc.zndsl"
```

melekes committed:
update test name

michalkucharczyk commented on zombienet/0004-webrtc/generate-warp-sync-database.toml:
```diff
@@ -0,0 +1,17 @@
+# this file is not intended to be executed in CI stage
```
If you are reusing database from `0001` you probably don't want to have it here. It is only used to (re-)generate database locally and update it on googlecloud storage.

michalkucharczyk commented on zombienet/0004-webrtc/test-webrtc.toml:
```diff
@@ -0,0 +1,30 @@
+  [[relaychain.nodes]]
+  name = "dave"
+  validator = false
+  args = ["--sync warp"]
```
Dumb question - I haven't reviewed entire PR:
How `webrtc` is enabled in this test?

michalkucharczyk replied:
I see: by default. So proably we don't need this test at all. `0001` should cover it.

melekes replied:
via chain spec. I've modified bootnodes URL to `/ip4/127.0.0.1/udp/30333/webrtc/p2p/12D3KooWFvMbTsNZ8peGS8dbnRvNDBspstupzwYC9NVwbzGCLtDt`

melekes replied:
> [0001](https://github.com/paritytech/substrate/blob/master/zombienet/0001-basic-warp-sync/test-warp-sync.toml) should cover it.

how? the default transport is still TCP. what I need is to disable TCP.. remove all protocols from advertised addresses except webrtc?

melekes committed:
remove generate-warp-sync-database from webrtc test

melekes committed:
change listen addr in zombienet tests

pepoviola commented on zombienet/0004-webrtc/test-webrtc.toml:
```diff
@@ -0,0 +1,33 @@
+  name = "alice"
+  validator = false
+  db_snapshot="https://storage.googleapis.com/zombienet-db-snaps/substrate/0001-basic-warp-sync/chains-0bb3f0be2ce41b5615b224215bcc8363aa0416a6.tgz"
+  args = ["--listen-addr /ip4/0.0.0.0/udp/30333/webrtc"]
```
Hi @melekes, this will require some changes in zombienet (and a new release). Let me work on that and I will update this pr.

Thanks!

pepoviola replied:
```suggestion
  args = ["--listen-addr", "/ip4/0.0.0.0/udp/30333/webrtc"]
```

pepoviola commented on zombienet/0004-webrtc/test-webrtc.toml:
```diff
@@ -0,0 +1,33 @@
+  name = "bob"
+  validator = false
+  db_snapshot="https://storage.googleapis.com/zombienet-db-snaps/substrate/0001-basic-warp-sync/chains-0bb3f0be2ce41b5615b224215bcc8363aa0416a6.tgz"
+  args = ["--listen-addr /ip4/0.0.0.0/udp/30333/webrtc"]
```
```suggestion
  args = ["--listen-addr",  "/ip4/0.0.0.0/udp/30333/webrtc"]
```

pepoviola commented on zombienet/0004-webrtc/test-webrtc.toml:
```diff
@@ -0,0 +1,33 @@
+  name = "charlie"
+  validator = false
+  db_snapshot="https://storage.googleapis.com/zombienet-db-snaps/substrate/0001-basic-warp-sync/chains-0bb3f0be2ce41b5615b224215bcc8363aa0416a6.tgz"
+  args = ["--listen-addr /ip4/0.0.0.0/udp/30333/webrtc"]
```
```suggestion
  args = ["--listen-addr",  "/ip4/0.0.0.0/udp/30333/webrtc"]
```

pepoviola commented on zombienet/0004-webrtc/test-webrtc.toml:
```diff
@@ -0,0 +1,33 @@
+  [[relaychain.nodes]]
+  name = "dave"
+  validator = false
+  args = ["--listen-addr /ip4/0.0.0.0/udp/30333/webrtc --sync warp"]
```
```suggestion
  args = ["--listen-addr",  "/ip4/0.0.0.0/udp/30333/webrtc",  --sync warp"]
```

pepoviola commented:
Hi @melekes, I make some suggestions in the zombienet files to make it works with the latest version (v1.3.30) that allow to override the `--listen-addr` flag. You will also need to update the version of zombienet in the `.gitlab-ci.yml` file (https://github.com/paritytech/substrate/blob/master/.gitlab-ci.yml#L58).

Thanks!!

melekes committed:
Apply suggestions from code review

Co-authored-by: Javier Viola <pepoviola@gmail.com>

melekes committed:
bump zombienet version

melekes commented:
Hmm... can't find my webrtc zombienet test

melekes committed:
Merge branch 'master' into anton/webrtc-transport

melekes commented:
> Hi @melekes, I make some suggestions in the zombienet files to make it works with the latest version (v1.3.30) that allow to override the --listen-addr flag. You will also need to update the version of zombienet in the .gitlab-ci.yml file (https://github.com/paritytech/substrate/blob/master/.gitlab-ci.yml#L58).

@pepoviola looks like the listen addr is still `/ws`: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2353638

pepoviola commented on .gitlab-ci.yml:
```diff
@@ -55,7 +55,7 @@ variables:
   NEXTEST_FAILURE_OUTPUT:          immediate-final
   NEXTEST_SUCCESS_OUTPUT:          final
-  ZOMBIENET_IMAGE:                 "docker.io/paritytech/zombienet:v1.3.22"
+  ZOMBIENET_IMAGE:                 "docker.io/paritytech/zombienet:v1.3.30"
```
```suggestion
  ZOMBIENET_IMAGE:                 "docker.io/paritytech/zombienet:v1.3.33"
```

Update the zombienet version to the latest.

pepoviola committed:
Update zombienet version.

melekes committed:
Merge branch 'master' into anton/webrtc-transport

melekes commented:
> > > Hi @melekes, I make some suggestions in the zombienet files to make it works with the latest version (v1.3.30) that allow to override the --listen-addr flag. You will also need to update the version of zombienet in the .gitlab-ci.yml file (https://github.com/paritytech/substrate/blob/master/.gitlab-ci.yml#L58).
> > 
> > 
> > @pepoviola looks like the listen addr is still `/ws`: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2353638
> 
> Hi @melekes, yes there was a missing change in that version. I updated this pr to the latest version of zombienet. Thanks for the feedback!!

Okay, great! However, now it fails w/ 

```
tar: can't create directory 'local_testnet/': Permission denied
tar: can't open 'local_testnet/': Is a directory
```

which looks like a problem with permissions somewhere...

pepoviola commented:
> local_testnet

Ahh yes, we should change the db snapshot to include the `data` directory. I will upload a new version of the same snap with the new directory format.

Thanks!!

pepoviola commented on zombienet/0004-webrtc/test-webrtc.toml:
```diff
@@ -0,0 +1,33 @@
+  [[relaychain.nodes]]
+  name = "alice"
+  validator = false
+  db_snapshot="https://storage.googleapis.com/zombienet-db-snaps/substrate/0001-basic-warp-sync/chains-0bb3f0be2ce41b5615b224215bcc8363aa0416a6.tgz"
```
```suggestion
  db_snapshot="https://storage.googleapis.com/zombienet-db-snaps/substrate/0001-basic-warp-sync/chains-a233bbacff650aac6bcb56b4e4ef7db7dc143cfb.tgz"
```

pepoviola commented on zombienet/0004-webrtc/test-webrtc.toml:
```diff
@@ -0,0 +1,33 @@
+  [[relaychain.nodes]]
+  name = "bob"
+  validator = false
+  db_snapshot="https://storage.googleapis.com/zombienet-db-snaps/substrate/0001-basic-warp-sync/chains-0bb3f0be2ce41b5615b224215bcc8363aa0416a6.tgz"
```
```suggestion
  db_snapshot="https://storage.googleapis.com/zombienet-db-snaps/substrate/0001-basic-warp-sync/chains-a233bbacff650aac6bcb56b4e4ef7db7dc143cfb.tgz"
```

pepoviola commented on zombienet/0004-webrtc/test-webrtc.toml:
```diff
@@ -0,0 +1,33 @@
+  [[relaychain.nodes]]
+  name = "charlie"
+  validator = false
+  db_snapshot="https://storage.googleapis.com/zombienet-db-snaps/substrate/0001-basic-warp-sync/chains-0bb3f0be2ce41b5615b224215bcc8363aa0416a6.tgz"
```
```suggestion
  db_snapshot="https://storage.googleapis.com/zombienet-db-snaps/substrate/0001-basic-warp-sync/chains-a233bbacff650aac6bcb56b4e4ef7db7dc143cfb.tgz"
```

pepoviola committed:
Update zombienet/0004-webrtc/test-webrtc.toml

pepoviola committed:
Update zombienet/0004-webrtc/test-webrtc.toml

pepoviola committed:
Update zombienet/0004-webrtc/test-webrtc.toml

pepoviola commented on .gitlab-ci.yml:
```diff
@@ -55,7 +55,7 @@ variables:
   NEXTEST_FAILURE_OUTPUT:          immediate-final
   NEXTEST_SUCCESS_OUTPUT:          final
-  ZOMBIENET_IMAGE:                 "docker.io/paritytech/zombienet:v1.3.22"
+  ZOMBIENET_IMAGE:                 "docker.io/paritytech/zombienet:v1.3.33"
```
```suggestion
  ZOMBIENET_IMAGE:                 "docker.io/paritytech/zombienet:v1.3.34"
```

pepoviola committed:
Update .gitlab-ci.yml

pepoviola commented:
Hi @melekes I updated the version of zombienet and the link to the new snap.
Thanks!

pepoviola committed:
bump version of zombienet and update snaps links (#13359)

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2378714

melekes committed:
Merge branch 'master' into anton/webrtc-transport

melekes commented:
@pepoviola thanks for helping me with this!

`--bootnodes /ip4/10.24.163.41/tcp/30333/ws/p2p/12D3KooWQCkBm1BYtkHpocxCwMgR8yjitEeHGx8spzcDLGt2gkBm`

Is there a way to change this to `--bootnodes /ip4/10.24.163.41/udp/30333/webrtc/p2p/12D3KooWQCkBm1BYtkHpocxCwMgR8yjitEeHGx8spzcDLGt2gkBm` ?

melekes commented:
@pepoviola friendly ping re https://github.com/paritytech/substrate/pull/12529#issuecomment-1427537060

pepoviola commented:
> @pepoviola friendly ping re [#12529 (comment)](https://github.com/paritytech/substrate/pull/12529#issuecomment-1427537060)

Hi @melekes sorry about the delay, let me check this in the current version of zombienet.
Thanks!

melekes committed:
Merge branch 'master' into anton/webrtc-transport

melekes committed:
bump zombienet version

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2500001

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2500052

melekes committed:
remove unneeded lines from zombienet test

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2500102

melekes committed:
add missing features

tomaka commented:
What is blocking this PR? To me it was ready already back in November.

melekes commented:
https://gitlab.parity.io/parity/mirrors/substrate/builds/2500337 is failing + https://github.com/libp2p/rust-libp2p/pull/3575 (I think we should wait for a new patch release)

altonen commented:
@tomaka 

There was a bug in libp2p (https://github.com/libp2p/rust-libp2p/issues/3244) which prevents `Identify` from identifying peers that advertise addresses it doesn't know about and would cause `PeerIdentify` in Substrate to fail for these peers. We made a decision to postpone releasing QUIC and WebRTC until the network has had some time to upgrade to the version of libp2p that includes a fix to this bug.

mxinden commented:
> + [libp2p/rust-libp2p#3575](https://github.com/libp2p/rust-libp2p/pull/3575) (I think we should wait for a new patch release)

Released as `libp2p` `v0.51.1` and `libp2p-webrtc` `v0.4.0-alpha.3`.

https://github.com/libp2p/rust-libp2p/blob/libp2p-v0.51.1/transports/webrtc/CHANGELOG.md#040-alpha3

> There was a bug in libp2p ([libp2p/rust-libp2p#3244](https://github.com/libp2p/rust-libp2p/issues/3244)) which prevents `Identify` from identifying peers that advertise addresses it doesn't know about and would cause `PeerIdentify` in Substrate to fail for these peers. We made a decision to postpone releasing QUIC and WebRTC until the network has had some time to upgrade to the version of libp2p that includes a fix to this bug.

Additional context here:

https://github.com/libp2p/rust-libp2p/blob/master/protocols/identify/CHANGELOG.md#0411

melekes committed:
Merge branch 'master' into anton/webrtc-transport

melekes committed:
enable logging for bob

melekes commented:
the zombienet tests are currently failing because the `bootnode` URL lacks `/certhash/XXX` component.

pepoviola commented:
> the zombienet tests are currently failing because the `bootnode` URL lacks `/certhash/XXX` component.



> the zombienet tests are currently failing because the `bootnode` URL lacks `/certhash/XXX` component.

Hi @melekes, we can add the missing component to the `bootnode` URL, currently we set something like this

```
/ip4/10.27.24.16/udp/30333/webrtc/p2p/12D3KooWQCkBm1BYtkHpocxCwMgR8yjitEeHGx8spzcDLGt2gkBm
```
Thanks!

melekes commented:
@pepoviola right, so I can add a flag similar to `--node-key`, but x509 certificate is pretty big...

it will be smth like

`--webrtc-certificate-raw '2d2d2d2d2d424547494e20455850495245532d2d2d2d2d0d0a415066686e6738414141413d0d0a2d2d2d2d2d454e4420455850495245532d2d2d2d2d0d0a0a2d2d2d2d2d424547494e20505249564154455f4b45592d2d2d2d2d0d0a4d494748416745414d424d4742797147534d34394167454743437147534d3439417745484247307761774942415151674663347473774a657230546e47524f710d0a45497a534b506a625a3937474149463078596d50516663563967576852414e4341415374386a4f6f42452f4847493751504f56354b6471575041644d525347340d0a3569374b524f536a754d5077493658524c7079782f71693770514548796378735059532f66454f543041444f644945594a532b6e364954530d0a2d2d2d2d2d454e4420505249564154455f4b45592d2d2d2d2d0d0a0d0a2d2d2d2d2d424547494e2043455254494649434154452d2d2d2d2d0d0a4d494942574443422f36414441674543416768616d33776c41587350707a414b42676771686b6a4f50515144416a41684d523877485159445651514444425a790d0a5932646c6269427a5a57786d49484e705a32356c5a43426a5a584a304d434158445463314d4445774d5441774d4441774d466f59447a51774f5459774d5441780d0a4d4441774d444177576a41684d523877485159445651514444425a795932646c6269427a5a57786d49484e705a32356c5a43426a5a584a304d466b77457759480d0a4b6f5a497a6a3043415159494b6f5a497a6a304441516344516741457266497a714152507878694f30447a6c65536e616c6a774854455568754f5975796b546b0d0a6f376a4438434f6c305336637366366f7536554242386e4d62443245763378446b3941417a6e534247435576702b694530714d664d42307747775944565230520d0a42425177456f49514f48686a61546471544441305a47523352446c6e527a414b42676771686b6a4f5051514441674e4941444246416942764f7a3942384353540d0a367248374a364175486a577a456c704662595536776779784b69544e6a7669345a51496841493374667931794e37734f5936764455594c71566c50494f376d470d0a58335845717766524d6843744b54464c0d0a2d2d2d2d2d454e442043455254494649434154452d2d2d2d2d0d0a'`

I guess there's no other way to have persisted webrtc certificate for the bootnode

melekes committed:
add `--webrtc-certificate-raw` flag for zombienet

melekes commented:
The one thing left is to modify zombienet again to allow custom protocols inside bootnode URL.

pepoviola commented:
> 2d2d2d2d2d424547494e20455850495245532d2d2d2d2d0d0a415066686e6738414141413d0d0a2d2d2d2d2d454e4420455850495245532d2d2d2d2d0d0a0a2d2d2d2d2d424547494e20505249564154455f4b45592d2d2d2d2d0d0a4d494748416745414d424d4742797147534d34394167454743437147534d3439417745484247307761774942415151674663347473774a657230546e47524f710d0a45497a534b506a625a3937474149463078596d50516663563967576852414e4341415374386a4f6f42452f4847493751504f56354b6471575041644d525347340d0a3569374b524f536a754d5077493658524c7079782f71693770514548796378735059532f66454f543041444f644945594a532b6e364954530d0a2d2d2d2d2d454e4420505249564154455f4b45592d2d2d2d2d0d0a0d0a2d2d2d2d2d424547494e2043455254494649434154452d2d2d2d2d0d0a4d494942574443422f36414441674543416768616d33776c41587350707a414b42676771686b6a4f50515144416a41684d523877485159445651514444425a790d0a5932646c6269427a5a57786d49484e705a32356c5a43426a5a584a304d434158445463314d4445774d5441774d4441774d466f59447a51774f5459774d5441780d0a4d4441774d444177576a41684d523877485159445651514444425a795932646c6269427a5a57786d49484e705a32356c5a43426a5a584a304d466b77457759480d0a4b6f5a497a6a3043415159494b6f5a497a6a304441516344516741457266497a714152507878694f30447a6c65536e616c6a774854455568754f5975796b546b0d0a6f376a4438434f6c305336637366366f7536554242386e4d62443245763378446b3941417a6e534247435576702b694530714d664d42307747775944565230520d0a42425177456f49514f48686a61546471544441305a47523352446c6e527a414b42676771686b6a4f5051514441674e4941444246416942764f7a3942384353540d0a367248374a364175486a577a456c704662595536776779784b69544e6a7669345a51496841493374667931794e37734f5936764455594c71566c50494f376d470d0a58335845717766524d6843744b54464c0d0a2d2d2d2d2d454e442043455254494649434154452d2d2d2d2d0d0a

Hi @melekes, as workaround I will add a new field (`node_cert_hash`) to be set at node configuration level (in zombienet) and been used to build the bootnode string.

Thanks!

pepoviola commented on .gitlab-ci.yml:
```diff
@@ -55,7 +55,7 @@ variables:
   NEXTEST_FAILURE_OUTPUT: immediate-final
   NEXTEST_SUCCESS_OUTPUT: final
-  ZOMBIENET_IMAGE: "docker.io/paritytech/zombienet:v1.3.37"
+  ZOMBIENET_IMAGE: "docker.io/paritytech/zombienet:v1.3.38"
```
```suggestion
  ZOMBIENET_IMAGE: "docker.io/paritytech/zombienet:v1.3.42"
```

pepoviola replied:
The latest version support set `p2p_cert_hash` for the node.

pepoviola commented on zombienet/0004-webrtc/test-webrtc.toml:
```diff
@@ -0,0 +1,33 @@
+
+  [[relaychain.nodes]]
+  name = "alice"
+  validator = false
```
```suggestion
  validator = false
  p2p_cert_hash = "uEiAvF2IuPZJuKkdb5E-gSM_AmumLi3P5ZR_fvZGPE3vJEA"
```

Hash of the cert used in this test.

pepoviola commented:
Hi @melekes, I add a couple of suggestions to use the latest version of zombienet that support set the `p2p_cert_hash` for the node. Let me know if you have questions.
Thanks!

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2575042

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2575041

melekes committed:
Merge branch 'master' into anton/webrtc-transport

melekes committed:
zombienet: set p2p_cert_hash

melekes committed:
set p2p_cert_hash for alice

melekes committed:
Merge branch 'master' into anton/webrtc-transport

melekes commented:
> Please make sure to use `multiaddr` `>= v0.17.1`. See [multiformats/rust-multiaddr#84](https://github.com/multiformats/rust-multiaddr/pull/84).

Thanks for mentioning 🙏  but we only depend on it indirectly via `rust-libp2p`.

melekes committed:
try to fix unsupported address error

```
m": Other(Custom { kind: Other, error: A(Other(A(A(A(A(B(MultiaddrNotSupported("/ip4/10.24.104.30/udp/30333/webrtc/certhash/uEiAvF2IuPZJuKkdb5E-gSM_AmumLi3P5ZR_fvZGPE3vJEA/p2p/12D3KooWQCkBm1BYtkHpocxCwMgR8yjitEeHGx8spzcDLGt2gkBm")))))))) })
```
by changing the ordering of transports

mxinden commented:
> > Please make sure to use `multiaddr` `>= v0.17.1`. See [multiformats/rust-multiaddr#84](https://github.com/multiformats/rust-multiaddr/pull/84).
> 
> Thanks for mentioning pray but we only depend on it indirectly via `rust-libp2p`.

Right. But rust-libp2p won't explicitly require `multiaddr` `>=0.17.1` any time soon, i.e. we might not cut a `libp2p` release in time. Thus you would need to ensure in your `Cargo.lock`, whether that file is here or in the Polkadot repository, that you are using `multiaddr` `>=0.17.1`. Does that make sense?

melekes commented:
> Does that make sense?

Yes.

jakehemmerle commented:
hey @melekes this is so close! are you able to finish? can you describe what is left?

melekes committed:
Merge branch 'master' into anton/webrtc-transport

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2918223

melekes committed:
remove leftovers from merge

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2918370

melekes committed:
add missing dev dep

melekes commented:
This should be good to go now 👍 

@paritytech/sdk-node would be good to get another approval before I merge the PR.

altonen commented:
Did you check how the Identify protocol works when the WebRTC listening endpoint is not specified? The node is able to establish connections? And when it's specified, the Identify fails, right?

Either way, please don't merge this yet.

melekes committed:
rename /webrtc to /webrtc-direct

see https://github.com/multiformats/rust-multiaddr/pull/84

melekes commented:
> And when it's specified, the Identify fails, right?

why do you expect it to fail?

melekes committed:
Merge branch 'master' into anton/webrtc-transport

tomaka commented:
Would it please be possible to take action here and finally deliver this? There has been almost no activity in the last 2 months. It feels like this PR has been put in a dusty corner, and occasionally the upper layer of dust gets removed.

bkchr commented:
This pr was depending on the libp2p update. This has been merged. @melekes job is now to verify that nodes are able to connect to main net, without having `webrtc` enabled. If that works, we can merge directly.

melekes commented:
There was also a bug in `libp2p-identify` - https://github.com/libp2p/rust-libp2p/issues/3244. The last polkadot release contains the patched version (https://github.com/paritytech/polkadot/blob/9b1fc27cec47f01a2c229532ee7ab79cc5bb28ef/Cargo.lock#L4018-L4019), so we should be good to go. I'm going to fix polkadot CI job and merge this.

melekes committed:
Revert "rename /webrtc to /webrtc-direct"

This reverts commit 2b37e04fa945d187d456838e6505a19bcfa8f6a5.

melekes commented:
someone from @paritytech/core-devs must approve this

bkchr commented:
> someone from @paritytech/core-devs must approve this

@altonen should do the final review

melekes commented:
> hey @melekes this is so close! are you able to finish? can you describe what is left?

https://github.com/paritytech/substrate/pull/12529#issuecomment-1576438086

altonen commented on client/network/src/transport.rs:
```diff
@@ -99,8 +102,22 @@ pub fn build_transport(
-		.boxed();
+		.timeout(Duration::from_secs(20));
+
+	if !memory_only {
```
I sent you a DM on Element but out of caution I think the WebRTC should be enabled only if the listening endpoint was provided because some of the network has not upgraded to 0.9.42 which should mean that Identify would fail for these peers as they don't recognize WebRTC. After a while we could remove this extra check and enable it by default.

tomaka replied:
I can read [here](https://github.com/paritytech/polkadot/releases) that Polkadot 0.9.42 is a "high priority" release, and node operators are supposed to upgrade as soon as possible. The release was a month ago.
My opinion is thus that it should be called out as their fault.

bkchr replied:
But there are also downstream users (parachains) which do not follow the same release schedule etc.

altonen replied:
That's a fair point, any thoughts @bkchr?

I looked at Moonbeam master and it looks like they're still on v0.9.40 which means this change could break their chain unless we upgrade libp2p-identify for v0.9.40. I'm not sure how that process goes though

melekes replied:
What if they never upgrade? We should enable webrtc for non-validators by default at some point. At least that was the plan in the beginning.

altonen replied:
I think they will eventually update and we can enable it for non-validators. AFAIR in December we discussed this and decided that waiting for 2-3 release should be enough

altonen replied:
The plan is to enable it for Westend at least and we can look into enabling it for Kusama/Polkadot as well

melekes committed:
Revert "Revert "rename /webrtc to /webrtc-direct""

This reverts commit 62a4658db091e9260da9c56bd3ca194e2d3a70ba.

melekes committed:
disable webrtc for non-validators

see https://github.com/paritytech/substrate/pull/12529#discussion_r1217863946

pepoviola commented:
> webrtc zombienet test failed again... not sure what changed since the last time this PR had been touched.

This are the [logs](https://grafana.parity-mgmt.parity.io/goto/Pex4UHlVg?orgId=1) from `Dave`

```


2023-06-05 10:37:05.276 | 2023-06-05 13:37:05 Looping and listening Ok(224.0.0.251:5353) |  
-- | -- | --
  |   | 2023-06-05 10:37:05.276 | 2023-06-05 13:37:05 signaling state changed to stable |  
  |   | 2023-06-05 10:37:05.276 | 2023-06-05 13:37:05 [controlling]: Setting new connection state: Checking |  
  |   | 2023-06-05 10:37:05.276 | 2023-06-05 13:37:05 [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet. |  
  |   | 2023-06-05 10:37:05.276 | 2023-06-05 13:37:05 [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet. |  
  |   | 2023-06-05 10:37:05.276 | 2023-06-05 13:37:05 ICE connection state changed: checking |  
  |   | 2023-06-05 10:37:05.480 | 2023-06-05 13:37:05 [controlling]: Failed to close candidate udp4 host 10.24.33.104:30333: the agent is closed |  
  |   | 2023-06-05 10:37:05.480 | 2023-06-05 13:37:05 [controlling]: Setting new connection state: Failed |  
  |   | 2023-06-05 10:37:05.480 | 2023-06-05 13:37:05 ICE connection state changed: failed |  
  |   | 2023-06-05 10:37:05.480 | 2023-06-05 13:37:05 peer connection state changed: failed |  
  |   | 2023-06-05 10:37:07.231 | 2023-06-05 13:37:07 ⏩ Warping, Downloading finality proofs, 0.00 Mib (0 peers), best: #0 (0xedb3…0fa9), finalized #0 (0xedb3…0fa9), ⬇ 0 ⬆ 0 |  
  |   | 2023-06-05 10:37:12.240 | 2023-06-05 13:37:12 ⏩ Warping, Downloading finality proofs, 0.00 Mib (0 peers), best: #0 (0xedb3…0fa9), finalized #0 (0xedb3…0fa9), ⬇ 0 ⬆ 0


```

sam0x17 commented on client/cli/src/params/webrtc_certificate_params.rs:
```diff
@@ -0,0 +1,172 @@
+		let params = WebRTCCertificateParams {
+			webrtc_certificate_file: None,
+			webrtc_certificate_ephemeral: None,
+			webrtc_certificate_raw: Some("2d2d2d2d2d424547494e20455850495245532d2d2d2d2d0d0a415066686e6738414141413d0d0a2d2d2d2d2d454e4420455850495245532d2d2d2d2d0d0a0a2d2d2d2d2d424547494e20505249564154455f4b45592d2d2d2d2d0d0a4d494748416745414d424d4742797147534d34394167454743437147534d3439417745484247307761774942415151674663347473774a657230546e47524f710d0a45497a534b506a625a3937474149463078596d505166635639 ...
```
```suggestion
			webrtc_certificate_raw: Some(String::from(
				"2d2d2d2d2d424547494e20455850495245532d2d2d2d2d0d0a415066686e6\
				738414141413d0d0a2d2d2d2d2d454e4420455850495245532d2d2d2d2d0d0\
				a0a2d2d2d2d2d424547494e20505249564154455f4b45592d2d2d2d2d0d0a4\
				d494748416745414d424d4742797147534d34394167454743437147534d343\
				9417745484247307761774942415151674663347473774a657230546e47524\
				f710d0a45497a534b506a625a3937474149463078596d50516663563967576\
				852414e4341415374386a4f6f42452f4847493751504f56354b64715750416\
				44d525347340d0a3569374b524f536a754d5077493658524c7079782f71693\
				770514548796378735059532f66454f543041444f644945594a532b6e36495\
				4530d0a2d2d2d2d2d454e4420505249564154455f4b45592d2d2d2d2d0d0a0\
				d0a2d2d2d2d2d424547494e2043455254494649434154452d2d2d2d2d0d0a4\
				d494942574443422f36414441674543416768616d33776c41587350707a414\
				b42676771686b6a4f50515144416a41684d523877485159445651514444425\
				a790d0a5932646c6269427a5a57786d49484e705a32356c5a43426a5a584a3\
				04d434158445463314d4445774d5441774d4441774d466f59447a51774f545\
				9774d5441780d0a4d4441774d444177576a41684d523877485159445651514\
				444425a795932646c6269427a5a57786d49484e705a32356c5a43426a5a584\
				a304d466b77457759480d0a4b6f5a497a6a3043415159494b6f5a497a6a304\
				441516344516741457266497a714152507878694f30447a6c65536e616c6a7\
				74854455568754f5975796b546b0d0a6f376a4438434f6c305336637366366\
				f7536554242386e4d62443245763378446b3941417a6e534247435576702b6\
				94530714d664d42307747775944565230520d0a42425177456f49514f48686\
				a61546471544441305a47523352446c6e527a414b42676771686b6a4f50515\
				14441674e4941444246416942764f7a3942384353540d0a367248374a36417\
				5486a577a456c704662595536776779784b69544e6a7669345a51496841493\
				374667931794e37734f5936764455594c71566c50494f376d470d0a5833584\
				5717766524d6843744b54464c0d0a2d2d2d2d2d454e4420434552544946494\
				34154452d2d2d2d2d0d0a"
			)),
```

sam0x17 replied:
btw it looks like this github just renders it badly:
```rust
			webrtc_certificate_raw: Some(String::from(
				"2d2d2d2d2d424547494e20455850495245532d2d2d2d2d0d0a415066686e6\
				738414141413d0d0a2d2d2d2d2d454e4420455850495245532d2d2d2d2d0d0\
				a0a2d2d2d2d2d424547494e20505249564154455f4b45592d2d2d2d2d0d0a4\
				d494748416745414d424d4742797147534d34394167454743437147534d343\
				9417745484247307761774942415151674663347473774a657230546e47524\
				f710d0a45497a534b506a625a3937474149463078596d50516663563967576\
				852414e4341415374386a4f6f42452f4847493751504f56354b64715750416\
				44d525347340d0a3569374b524f536a754d5077493658524c7079782f71693\
				770514548796378735059532f66454f543041444f644945594a532b6e36495\
				4530d0a2d2d2d2d2d454e4420505249564154455f4b45592d2d2d2d2d0d0a0\
				d0a2d2d2d2d2d424547494e2043455254494649434154452d2d2d2d2d0d0a4\
				d494942574443422f36414441674543416768616d33776c41587350707a414\
				b42676771686b6a4f50515144416a41684d523877485159445651514444425\
				a790d0a5932646c6269427a5a57786d49484e705a32356c5a43426a5a584a3\
				04d434158445463314d4445774d5441774d4441774d466f59447a51774f545\
				9774d5441780d0a4d4441774d444177576a41684d523877485159445651514\
				444425a795932646c6269427a5a57786d49484e705a32356c5a43426a5a584\
				a304d466b77457759480d0a4b6f5a497a6a3043415159494b6f5a497a6a304\
				441516344516741457266497a714152507878694f30447a6c65536e616c6a7\
				74854455568754f5975796b546b0d0a6f376a4438434f6c305336637366366\
				f7536554242386e4d62443245763378446b3941417a6e534247435576702b6\
				94530714d664d42307747775944565230520d0a42425177456f49514f48686\
				a61546471544441305a47523352446c6e527a414b42676771686b6a4f50515\
				14441674e4941444246416942764f7a3942384353540d0a367248374a36417\
				5486a577a456c704662595536776779784b69544e6a7669345a51496841493\
				374667931794e37734f5936764455594c71566c50494f376d470d0a5833584\
				5717766524d6843744b54464c0d0a2d2d2d2d2d454e4420434552544946494\
				34154452d2d2d2d2d0d0a"
			)),
```

melekes committed:
add logging to all nodes

tomaka commented:
If I understand [this comment](https://github.com/libp2p/rust-libp2p/issues/4049#issuecomment-1588830864) correctly, this feature is going to be blocked for another year or so, the time for rust-libp2p to rewrite its implementation from scratch?

melekes commented:
> If I understand https://github.com/libp2p/rust-libp2p/issues/4049#issuecomment-1588830864 correctly, this feature is going to be blocked for another year or so, the time for rust-libp2p to rewrite its implementation from scratch?

There's no need in rewrite right now. I.e. it doesn't block us from merging this. Only one bug remains to be fixed.

melekes committed:
Merge branch 'master' into anton/webrtc-transport

melekes committed:
trace webrtc package

seunlanlege commented:
Bump

melekes committed:
Merge branch 'master' into anton/webrtc-transport

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/infrastructure-treasury-funding/3507/3

melekes committed:
fixes after merge

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3257718

melekes committed:
fix protocol name

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3257797

melekes committed:
use libp2p-webrtc dep in CLI

melekes committed:
Merge branch 'master' into anton/webrtc-transport

melekes committed:
remove webrtc_sctp trace from dave's log

melekes committed:
bump regex to resolve polkadot companion CI job

```
[2023-07-26 09:13:50] error: failed to select a version for `regex`.
[2023-07-26 09:13:50]     ... required by package `webrtc v0.8.0`
[2023-07-26 09:13:50]     ... which satisfies dependency `webrtc = "^0.8.0"` of package `libp2p-webrtc v0.6.0-alpha`
[2023-07-26 09:13:50]     ... which satisfies dependency `libp2p-webrtc = "^0.6.0-alpha"` of package `sc-network v0.10.0-dev (/builds/parity/mirrors/substrate/client/network)`
[2023-07-26 09:13:50]     ... which satisfies git dependency `sc-network` (locked to 0.10.0-dev) of package `polkadot-availability-distribution v0.9.43 (/builds/parity/mirrors/substrate/companions/polkadot/node/network/availability-distribution)`
[2023-07-26 09:13:50]     ... which satisfies path dependency `polkadot-availability-distribution` (locked to 0.9.43) of package `polkadot-service v0.9.43 (/builds/parity/mirrors/substrate/companions/polkadot/node/service)`
[2023-07-26 09:13:50]     ... which satisfies path dependency `service` (locked to 0.9.43) of package `polkadot-cli v0.9.43 (/builds/parity/mirrors/substrate/companions/polkadot/cli)`
[2023-07-26 09:13:50]     ... which satisfies path dependency `polkadot-cli` (locked to 0.9.43) of package `polkadot v0.9.43 (/builds/parity/mirrors/substrate/companions/polkadot)`
[2023-07-26 09:13:50] versions that meet the requirements `^1.7.1` are: 1.9.1, 1.9.0, 1.8.4, 1.8.3, 1.8.2, 1.8.1, 1.8.0, 1.7.3, 1.7.2, 1.7.1
[2023-07-26 09:13:50]
[2023-07-26 09:13:50] all possible versions conflict with previously selected packages.
[2023-07-26 09:13:50]
[2023-07-26 09:13:50]   previously selected package `regex v1.6.0`
[2023-07-26 09:13:50]     ... which satisfies dependency `regex = "^1.1"` (locked to 1.6.0) of package `Inflector v0.11.4`
[2023-07-26 09:13:50]     ... which satisfies dependency `Inflector = "^0.11.4"` (locked to 0.11.4) of package `frame-benchmarking-cli v4.0.0-dev (/builds/parity/mirrors/substrate/utils/frame/benchmarking-cli)`
[2023-07-26 09:13:50]     ... which satisfies git dependency `frame-benchmarking-cli` (locked to 4.0.0-dev) of package `polkadot-cli v0.9.43 (/builds/parity/mirrors/substrate/companions/polkadot/cli)`
[2023-07-26 09:13:50]     ... which satisfies path dependency `polkadot-cli` (locked to 0.9.43) of package `polkadot v0.9.43 (/builds/parity/mirrors/substrate/companions/polkadot)`
[2023-07-26 09:13:50]
[2023-07-26 09:13:50] failed to select a version for `regex` which could resolve this conflict
```

melekes committed:
more logging

mxinden commented:
> zombienet test failed again, but with different error this time:
> 
> ```
> 2023-07-25 16:48:35.169  INFO tokio-runtime-worker libp2p_kad::handler: Remote 12D3KooWQCkBm1BYtkHpocxCwMgR8yjitEeHGx8spzcDLGt2gkBm no longer supports our kademlia protocol
> 
> 2023-07-25 16:48:35.169 TRACE tokio-runtime-worker sub-libp2p: Ignoring self-reported address /ip4/10.24.122.238/udp/30333/webrtc-direct/certhash/uEiAvF2IuPZJuKkdb5E-gSM_AmumLi3P5ZR_fvZGPE3vJEA from 12D3KooWQCkBm1BYtkHpocxCwMgR8yjitEeHGx8spzcDLGt2gkBm as remote node is not part of the Kademlia DHT supported by the local node.
> 
> 2023-07-25 16:48:37.548 TRACE tokio-runtime-worker sub-libp2p: Libp2p => Query for 0020cc78cfbc884fd5668ebb8f2948752f2dbc3a4c665b1b678f6190ec7208cbff72 yielded 0 results    
> 2023-07-25 16:48:37.548 DEBUG tokio-runtime-worker sub-libp2p: Libp2p => Random Kademlia query has yielded empty results
> ```
> 
> not yet sure what is happening, but it looks like Kademlia does not provide dave with addresses to connect to...

Just an idea: are you either calling `Kademlia::set_mode(Mode::Server)` or are you providing an external address @melekes?

melekes commented:
> Just an idea: are you either calling `Kademlia::set_mode(Mode::Server)` or are you providing an external address @melekes?

@mxinden neither? I was hoping that the node will discover it's external address via bootnode and switch to server mode automatically.

melekes committed:
set mode to server by default

refs https://github.com/paritytech/substrate/pull/12529/#issuecomment-1651666805

melekes committed:
Revert "bump regex to resolve polkadot companion CI job"

This reverts commit 142b8849db52117469da003404a51ee39e7e2c72.

melekes committed:
Revert "set mode to server by default"

This reverts commit 7d4e8e28026feb05b4958f6ed5d66db4f4f1e63c.

melekes commented:
```
dave.log
5401:2023-07-26 12:24:15.170 DEBUG tokio-runtime-worker sub-libp2p: Libp2p => Connected(PeerId("12D3KooWPKzmmE2uYgF3z13xjpbFTp63g9dZFag8pG6MgnpSLF4S")) with errors: [("/ip4/10.24.11.252/udp/30333/webrtc-direct/p2p/12D3KooWPKzmmE2uYgF3z13xjpbFTp63g9dZFag8pG6MgnpSLF4S", Other(Custom { kind: Other, error: Right(Other(Left(Left(Left(Left(Right(MultiaddrNotSupported("/ip4/10.24.11.252/udp/30333/webrtc-direct/p2p/12D3KooWPKzmmE2uYgF3z13xjpbFTp63g9dZFag8pG6MgnpSLF4S")))))))) }))]    
5697:2023-07-26 12:24:15.176 DEBUG tokio-runtime-worker sub-libp2p: Libp2p => Connected(PeerId("12D3KooWRkZhiRhsqmrQ28rt73K7V3aCBpqKrLGSXmZ99PTcTZby")) with errors: [("/ip4/10.24.10.241/udp/30333/webrtc-direct/p2p/12D3KooWRkZhiRhsqmrQ28rt73K7V3aCBpqKrLGSXmZ99PTcTZby", Other(Custom { kind: Other, error: Right(Other(Left(Left(Left(Left(Right(MultiaddrNotSupported("/ip4/10.24.10.241/udp/30333/webrtc-direct/p2p/12D3KooWRkZhiRhsqmrQ28rt73K7V3aCBpqKrLGSXmZ99PTcTZby")))))))) }))]
```

looks like Dave is receiving multiaddresses without a `certhash` component

melekes committed:
Merge branch 'master' into anton/webrtc-transport

melekes committed:
reduce logging

melekes committed:
Merge branch 'master' into anton/webrtc-transport

melekes committed:
Merge branch 'master' into anton/webrtc-transport

melekes commented:
```
2023-07-31 09:04:31 Substream 10786 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10788 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10790 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10792 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10794 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10796 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10798 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10800 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10802 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10804 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10806 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10808 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10810 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10812 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10814 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10816 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10818 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10820 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10822 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10824 dropped without graceful close, sending Reset
2023-07-31 09:04:31 Substream 10826 dropped without graceful close, sending Reset
```

Not sure why Charlie needs 10000 substreams unless there's a bug somewhere.

melekes committed:
Merge branch 'master' into anton/webrtc-transport

altonen commented:
@melekes are these problems reproducible with just libp2p or is Substrate somehow causing these issues?

melekes committed:
more logging for dave
